main/src/ca/uwaterloo/flix/language/phase/Resolver.scala

```diff
@@ -261,60 +258,72 @@ object Resolver extends Phase[NamedAst.Root, ResolvedAst.Root] {
+        */
+      def visitSig(sig: NamedAst.Sig, tvar: Type.Var, loc: SourceLocation): ResolvedAst.Expression = {
+        // Find the arity of the function definition.
+        val arity = sig.fparams.length
 
-              // Create the fresh fparams
-              val fparams = mkFreshFparams(arity, loc)
+        // Create the fresh fparams
+        val fparams = mkFreshFparams(arity, loc)
 
-              // The definition expression.
-              val defExp = ResolvedAst.Expression.Def(defn.sym, tvar, loc)
+        // The signature expression.
+        val sigExp = ResolvedAst.Expression.Sig(sig.sym, tvar, loc)
 
-              // Create and apply the lambda expressions
-              mkCurriedLambda(fparams, defExp, loc)
+        // Create and apply the lambda expressions
+        mkCurriedLambda(fparams, sigExp, loc)
+      }
+
+      /**
+        * Resolve the application expression, performing currying over the subexpressions.
+        */
+      def visitApply(exp: NamedAst.Expression.Apply): Validation[ResolvedAst.Expression, ResolutionError] = exp match {
+        case NamedAst.Expression.Apply(exp, exps, loc) =>
+          for {
+            e <- visit(exp, tenv0)
+            es <- traverse(exps)(visit(_, tenv0))
+          } yield {
+            es.foldLeft(e) {
+              case (acc, a) => ResolvedAst.Expression.Apply(acc, List(a), Type.freshVar(Kind.Star), Type.freshVar(Kind.Bool), loc)
+            }
           }
+      }
 
-        case NamedAst.Expression.Sig(qname, tvar, loc) =>
-          lookupSig(qname, ns0, sigs) map {
-            case sig =>
-              //
-              // We must curry the definition. Otherwise we would not be here.
-              //
-              // We introduce /n/ lambda expressions around the definition expression.
-              //
 
-              // Find the arity of the function definition.
-              val arity = sig.fparams.length
+      /**
+        * Local visitor.
+        */
+      def visit(e0: NamedAst.Expression, tenv0: Map[Symbol.VarSym, Type]): Validation[ResolvedAst.Expression, ResolutionError] = e0 match {
 
-              // Create the fresh fparams
-              val fparams = mkFreshFparams(arity, loc)
+        case NamedAst.Expression.Wild(tvar, loc) =>
+          ResolvedAst.Expression.Wild(tvar, loc).toSuccess
 
-              // The definition expression.
-              val defExp = ResolvedAst.Expression.Sig(sig.sym, tvar, loc)
+        case NamedAst.Expression.Var(sym, loc) => tenv0.get(sym) match {
+          case None => ResolvedAst.Expression.Var(sym, sym.tvar, loc).toSuccess
+          case Some(tpe) => ResolvedAst.Expression.Var(sym, tpe, loc).toSuccess
+        }
 
-              // Create and apply the lambda expressions
-              mkCurriedLambda(fparams, defExp, loc)
+        case NamedAst.Expression.DefOrSig(qname, tvar, loc) =>
+          mapN(lookupDefSig(qname, ns0, root, sigs)) {
+            case Left(defn) => visitDef(defn, tvar, loc)
```

Comment:
I don't like the use of Left/Right. We should have a specialized type. (In other places we use something called LookupResult)